### DISCUSSION TITLE
How to point the build at the right ninja binary?

### DISCUSSION THREAD

**Discussion by rnikander** [android/ndk#2001] [Q&A] [answered]
I'm following the instructions [here](https://developer.android.com/studio/projects/install-ndk#vanilla_cmake), so I can use a newer version of CMake. There are 4 steps there, and 1-3 seem to have worked smoothly. However, at step 4, when I install a newer copy of Ninja (which I need for C++20 modules to work), the NDK build process still uses the version of `ninja.exe` in its directories. 

The build output contains:

```text
C++ build system [configure] failed while executing:
   @echo off
   "C:\\Program Files\\CMake\\bin\\cmake.exe" ^
   ...
   "-DCMAKE_MAKE_PROGRAM=C:\\Users\\rob\\AppData\\Local\\Android\\Sdk\\cmake\\3.22.1\\bin\\ninja.exe" ^
   ...
The Ninja generator does not support C++20 modules using Ninja version
    1.10.2
```
My ninja is at `C:\Users\rob\Workspace\Tools\ninja-win\ninja.exe`. It's on my path, and I have restarted Android Studio.

```text
PS C:\Users\rob> ninja --version
1.11.1
```
Is there something else I can do? Maybe it should be added to that 4th point in the documentation?

**Comment by DanAlbert** [comment, marked as answer]
As best as I can tell that's set [here](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:build-system/gradle-core/src/main/java/com/android/build/gradle/internal/cxx/configure/NativeLocationsBuildService.kt;l=191;drc=b5516899015633c99dc64b510d9729c4e001e89c) (I don't know how this works, I just went to the Studio section of cs.android.com and searched for `DCMAKE_MAKE_PROGRAM`). That seems to eventually call [`findNinjaPathLogic`](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:build-system/gradle-core/src/main/java/com/android/build/gradle/internal/cxx/configure/NinjaLocator.kt;l=27-32;drc=ba06981d00fe009bd0597c6131de61892408a25f), whose docs say:

```kotlin
/**
 * Method for locating ninja.exe. The search order is:
 * 1) Folder of CMake.exe, if present
 * 2) Other CMake SDK folders, if present
 * 3) Environment PATH
 */
```

So the one in your path is only used if you have no other CMake versions installed by the SDK manager. 2 and 3 seem backward to me, but that's another product. File a C++ bug against Studio at https://b.android.com if you want the behavior changed. It's pretty odd that there's not gradle DSL to override that.

**Reply by rnikander** [reply]
Reported [here](https://issuetracker.google.com/issues/326411158?pli=1). The workaround is easy enough. Just copy the `ninja.exe` to the one the SDK is using. 

But, C++20 modules still won't work, because the NDK does not include `clang-scan-deps`. I see it #1853 as a "feature request". I think that should be considered a bug, not a feature request. Hopefully it's an easy fix, to include the file that is built alongside `clang`.
